content/blog/2022-12-01-epidata-v4.Rmd

```diff
@@ -0,0 +1,128 @@
+
+We chose to switch from [MariaDB](https://mariadb.org) to [Percona Server for MySQL](https://www.percona.com/software/mysql-database/percona-server) for our database engine and management system.  Both are variants of the free open source [MySQL](https://www.mysql.com) database, and both provide additional support and features on top of the base product.  Percona offered us access to a more consistently developed drop-in replacement for MySQL, along with expertise and knowledge concentrated in both the open source community and Percona the company. Additionally, Percona Server for MySQL is a reliable target for monitoring with Percona’s [PMM](https://docs.percona.com/percona-monitoring-and-management/).
+
+To reduce our data footprint and speed lookups, we took steps to "[normalize](https://en.wikipedia.org/wiki/Database_normalization)" our database.  This involved redefining our schema to move some columns out of the main table and into tables of their own, which we reference with "[foreign key](https://en.wikipedia.org/wiki/Foreign_key)" columns instead.  By arranging our data this way, the long and often-repeated text strings that name our sources and signals as well as the geographic locations we cover are only stored once each, and referred to by a number instead.  The relatively short numbers take up much less space on disk compared to the text that they replace, and the database can optimize locating records by these numbers more easily than it can for text.  We employ two of these dimension tables, one for source and signal, and one for geographic type and value.  Using dimension tables efficiently requires [JOIN](https://en.wikipedia.org/wiki/Join_(SQL)) operations when retrieving from the database.  We avoided adding these slightly-more-complicated JOIN statements to many places in our code base by hiding them behind [VIEWs](https://en.wikipedia.org/wiki/View_(SQL)) -- our VIEWs are a shortcut to the JOIN logic that make our new normalized tables look virtually indistinguishable from old monolithic table to our web server code.
+
+We also added a "latest" table that includes a copy of just the issues that are most recent for all of our data points and removed the "latest issue" flag.  Since this no longer relies on looking in our gigantic collection of all issues for data that is flagged as being the latest, we gained a significant improvement to query times.  This is the data that drives our main [website](https://delphi.cmu.edu/covidcast/), so it is important for these operations to be fast for those browsing, and since we have so many visitors, it adds up to reduce our load significantly.  The "latest table" simplifies our data management responsibilities compared to the "latest flag", in that we don't have to clear a flag from an old row (a previously "latest" issue of data) when a new "latest" row is added -- instead we simply overwrite the old row in our latest table, and the database does this with great ease.
```

Comment:
> We also added a "latest" table that includes a copy of just the issues that are most recent for all of our data points and removed the **"latest issue"** flag.

Should we use the same language here as in the _2. Data Ingestion/Repair and Development_ section? ("is_latest flag")

Reply:
`"latest issue"` also appears elsewhere in the document (line 61 above).  we could do s/"latest issue"/is_latest/ here, but i think it looks a little silly without standing out in a fixed-width font or similar.  anyone reading and actually paying attention should be able to recognize that this "latest issue" flag is the same as the is_latest flag.  TL;DR: probbly leave it as-is

Reply:
Added fixed-width font, change a bit of wording to show both "latest issue" and ``is_latest``
<img width="747" alt="Screenshot 2023-01-25 at 9 53 54 AM" src="https://user-images.githubusercontent.com/118945681/214595941-d169dd74-2512-4765-acd9-cdc86b1a5f9e.png">
<img width="726" alt="Screenshot 2023-01-25 at 9 54 23 AM" src="https://user-images.githubusercontent.com/118945681/214596053-16f68f4b-6f10-4c02-af10-47967da71c00.png">